server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamRoute.java

```diff
@@ -94,31 +99,29 @@ public Upstream get() {
      * @param status - response http status; typically, 5xx or 429
      * @param retryAfterSeconds - the amount of seconds after which upstream should be available; if status 5xx this value ignored
      */
-    public void fail(HttpStatus status, long retryAfterSeconds) {
-        if (upstreamState != null) {
-            upstreamState.failed(status, retryAfterSeconds);
-        }
+    void fail(HttpStatus status, long retryAfterSeconds) {
+        balancer.fail(upstream, status, retryAfterSeconds);
```

Comment:
Make sure that API is used as expected with Objects.requireNonNull(upstream)

Reply:
fixed